Mathlib/CategoryTheory/Monoidal/Comon_.lean

```diff
@@ -0,0 +1,173 @@
+
+attribute [reassoc] Comon_.counit_comul Comon_.comul_counit
+
+-- We prove a more general `@[simp]` lemma below.
+attribute [reassoc (attr := simp)] Comon_.comul_assoc
+
+namespace Comon_
+
+/-- The trivial comonoid object. We later show this is initial in `Comon_ C`.
+-/
+@[simps]
+def trivial : Comon_ C where
+  X := 𝟙_ C
+  counit := 𝟙 _
+  comul := (λ_ _).inv
+  comul_assoc := by coherence
+  counit_comul := by coherence
+  comul_counit := by coherence
+
+instance : Inhabited (Comon_ C) :=
+  ⟨trivial C⟩
+
+variable {C}
+variable {M : Comon_ C}
+
+@[simp]
+theorem counit_comul_hom {Z : C} (f : M.X ⟶ Z) : M.comul ≫ (M.counit ⊗ f) = f ≫ (λ_ Z).inv := by
+  rw [leftUnitor_inv_naturality, tensorHom_def, counit_comul_assoc]
+
+@[simp]
+theorem mul_one_hom {Z : C} (f : M.X ⟶ Z) : M.comul ≫ (f ⊗ M.counit) = f ≫ (ρ_ Z).inv := by
+  rw [rightUnitor_inv_naturality, tensorHom_def', comul_counit_assoc]
+
+theorem assoc_flip :
+    M.comul ≫ (𝟙 M.X ⊗ M.comul) = M.comul ≫ (M.comul ⊗ 𝟙 M.X) ≫ (α_ M.X M.X M.X).hom := by simp
+
+/-- A morphism of comonoid objects. -/
+@[ext]
+structure Hom (M N : Comon_ C) where
+  /-- The underlying morphism of a morphism of comonoid objects. -/
+  hom : M.X ⟶ N.X
+  hom_counit : hom ≫ N.counit = M.counit := by aesop_cat
+  hom_comul : hom ≫ N.comul = M.comul ≫ (hom ⊗ hom) := by aesop_cat
+
+attribute [reassoc (attr := simp)] Hom.hom_counit Hom.hom_comul
+
+/-- The identity morphism on a comonoid object. -/
+@[simps]
+def id (M : Comon_ C) : Hom M M where
+  hom := 𝟙 M.X
+
+instance homInhabited (M : Comon_ C) : Inhabited (Hom M M) :=
+  ⟨id M⟩
+
+/-- Composition of morphisms of monoid objects. -/
+@[simps]
+def comp {M N O : Comon_ C} (f : Hom M N) (g : Hom N O) : Hom M O where
+  hom := f.hom ≫ g.hom
+
+instance : Category (Comon_ C) where
+  Hom M N := Hom M N
+  id := id
+  comp f g := comp f g
+
+@[ext] lemma ext {X Y : Comon_ C} {f g : X ⟶ Y} (w : f.hom = g.hom) : f = g := Hom.ext _ _ w
+
+@[simp] theorem id_hom' (M : Comon_ C) : (𝟙 M : Hom M M).hom = 𝟙 M.X := rfl
+
+@[simp]
+theorem comp_hom' {M N K : Comon_ C} (f : M ⟶ N) (g : N ⟶ K) :
+    (f ≫ g : Hom M K).hom = f.hom ≫ g.hom :=
+  rfl
+
+section
+
+variable (C)
+
+/-- The forgetful functor from comonoid objects to the ambient category. -/
+@[simps]
+def forget : Comon_ C ⥤ C where
+  obj A := A.X
+  map f := f.hom
+
+end
+
+instance forget_faithful : Faithful (@forget C _ _) where
+
+instance {A B : Comon_ C} (f : A ⟶ B) [e : IsIso ((forget C).map f)] : IsIso f.hom := e
+
+/-- The forgetful functor from comonoid objects to the ambient category reflects isomorphisms. -/
+instance : ReflectsIsomorphisms (forget C) where
+  reflects f e :=
+    ⟨⟨{ hom := inv f.hom }, by aesop_cat⟩⟩
+
+/-- Construct an isomorphism of monoids by giving an isomorphism between the underlying objects
+and checking compatibility with unit and multiplication only in the forward direction.
+-/
+def isoOfIso {M N : Comon_ C} (f : M.X ≅ N.X) (f_counit : f.hom ≫ N.counit = M.counit)
```

Comment:
We're quite inconstent about the naming of these definitions. In comma categories, they're called `isoMk`, for cones it's called `ext`, etc.